07-errors.md

```diff
@@ -0,0 +1,311 @@
+successfully.
+
+All non-zero exit codes indicate that the message _was not_ applied and that the chain state _was
+not_ updated.
+
+### System Non-Zero Exit Codes
+
+System non-zero exit codes are generated by the VM itself to indicate that the message itself failed
+to apply. Actors may not call `vm::abort` with any system exit code.
+
+| Code  | Status                     | Description                                                |
+|-------|----------------------------|------------------------------------------------------------|
+| 1     | `SysErrSenderInvalid`      | sender doesn't exist                                       |
+| 2     | `SysErrSenderStateInvalid` | message sender is not in a valid state to send the message |
+| 3     | reserved                   |                                                            |
+| 4     | `SysErrIllegalInstruction` | message receiver trapped                                   |
+| 5     | `SysErrInvalidReceiver`    | receiver doesn't exist and can't be automatically created  |
+| 6     | `SysErrInsufficientFunds`  | sender didn't have the requisite funds                     |
+| 7     | `SysErrOutOfGas`           | out of gas                                                 |
+| 8     | reserved                   |                                                            |
+| 9     | `SysErrIllegalExitCode`    | message receiver aborted with an system exit code          |
+| 10    | `SysErrAssertionFailed`    | some internal assertion failed                             |
+| 11-15 | reserved                   |                                                            |
+
+Notes:
+
+1. If a message sender doesn't exist on-chain, the VM will return a receipt with the
+   `ExitCode::SysErrSenderInvalid` exit code.
+2. If an actor calls `vm::abort` with a system exit code, that exit code will be replaced with
+   `ExitCode::SysErrIllegalExitCode`.
+3. If an actor panics/crashes/etc., the exit code `ExitCode::SysErrIllegalInstruction` will be
+   returned.
+4. If a message runs out of gas, the VM will return a receipt with the `ExitCode::SysErrOutOfGas`
+   exit code.
+5. Most of these exit codes are only reflected on-chain and will not be observed in internal sends.
+   The exceptions are:
+    - `SysErrIllegalExitCode` if the called actor aborted with a system exit code.
+    - `SysErrIllegalInstruction` if the called actor paniced.
+
+Changes from pre-FVM Filecoin:
+
+1. Removed `ExitCode::SysErrInvalidMethod` (3). Dispatch is now handled inside actor code.
+2. Removed `ExitCode::SysErrForbidden` (8). Caller verification is a userspace operation, we'll now
+   exit with `ExitCode::ErrForbidden`.
+4. Replaced `ExitCode::SysErrIllegalArgument` with `ExitCode::SysErrAssertionFailed` to more
+   accurately reflect its usage. When this appears on-chain, it means that a message hit some
+   internal Filecoin assert that shouldn't happen.
+
+### Standard Non-Zero Exit  Codes
```

Comment:
nit: would suggest using the term _ordinary_ instead of _standard_ here. Rationale: this whole document defines the _standards_ for error conditions in the Filecoin protocol, so it's a bit ambiguous because from that viewpoint, one could argue that System exit codes are standard (as in: defined by the standards).

```suggestion
### Ordinary Non-Zero Exit  Codes
```

Reply:
Ordinary isn't the right word. I'd go with either standardized and/or well-known (which is what it was before).

Reply:
Or just "standard user non-zero exit codes". I.e.:

1. System exit codes.
2. User exit codes.
    1. Standard user exit codes.
    2. Everything else.

Reply:
Let's punt this to a new PR.